File: Python/InsertionSort.py
import time

# Average complexity: O(n^2)
# Worst case complexity: O(n^2)

def insertion_sort(arr):
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i - 1
        while j >= 0 and key < arr[j]:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key
    return arr

# numbers are separated by spaces
def read_dataset():
    file = '../Dataset/decreasing_1M.txt'
    with open(file, 'r') as f:
        data = f.read().split()
    return list(map(int, data))

def main():
    arr = read_dataset()
    print('Unsorted array:', arr[:10], '...')
    start_time = time.perf_counter()
    arr = insertion_sort(arr)
    end_time = time.perf_counter()
    execution_time_ms = int((end_time - start_time) * 1000)
    print("Execution Time in Milliseconds:", execution_time_ms)
    print('Sorted array:', arr[:10], '...')

File: Java/InsertionSort.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.io.File;
import java.util.stream.Collectors;
import java.io.FileNotFoundException;

// Average complexity: O(n^2)
// Worst case complexity: O(n^2)

public class InsertionSort {

    private static final int MAX_SIZE = 1000000;

    public static void main(String args[]) {
        String fileName = "dataset_95_sorted.txt";
        if (args.length != 0) {
            fileName = args[0];
        }

        try {
            int[] dataset = readDataset(fileName);

            long startTime = System.currentTimeMillis();
            insertionSort(dataset);
            long endTime = System.currentTimeMillis();

            long elapsedTime = endTime - startTime;

            System.out.println("Elapsed time: " + elapsedTime + " miliseconds");
        } catch (FileNotFoundException e) {
            System.out.println("An error occurred.");
            e.printStackTrace();
        }
    }

    // Read txt file (integers separated by new space)
    public static int[] readDataset(String filename) throws FileNotFoundException {
        File file = new File(filename);
        Scanner scanner = new Scanner(file);
        List<Integer> numbers = new ArrayList<>();
        while (scanner.hasNext()) {
            if (numbers.size() == MAX_SIZE) {
                break;
            }
            numbers.add(scanner.nextInt());
        }
        scanner.close();
        return numbers.stream().mapToInt(i -> i).toArray();
    }

    public static void insertionSort(int[] numbers) {
        for (int currentIndex = 1; currentIndex < numbers.length; currentIndex++) {
            int currentNumber = numbers[currentIndex];
            int previousIndex = currentIndex - 1;
            while (previousIndex >= 0 && numbers[previousIndex] > currentNumber) {
                numbers[previousIndex + 1] = numbers[previousIndex];
                previousIndex = previousIndex - 1;
            }
            numbers[previousIndex + 1] = currentNumber;
        }
    }
}